### ISSUE TITLE
Preview HTTP and HTTPS URLs

### ISSUE THREAD

**Issue by PanderMusubi** [siacs/Conversations#1778]
Please, offer preview HTTP and HTTPS URLs in similar way WhatsApp and Telegram does. Make this an optional feature.

**Comment by licaon-kter** [comment]
This can raise some privacy issues.

**Comment by PanderMusubi** [comment]
Therefore should it be an option which is off by default. Perhaps a small warning should be shown when it is enabled or near the setting of the option.

**Comment by iNPUTmice** [comment]
I'm neither using WhatsApp nor Telegram so I don't know what you mean. Can you either describe this in more detail or illustrate with a screenshot what you mean.

**Comment by licaon-kter** [comment]
https://telegram.org/blog/link-preview

**Comment by PanderMusubi** [comment]
It would only apply to the first url with HTTP or HTTPS that is found in the message. If there is an error, display nothing.

**Comment by iNPUTmice** [comment]
Usually this is something that is being handled on the server side. Currently there is no XEP for this.

Better resources: http://ogp.me and http://telegram.wiki/tips:linkpreview

**Comment by PanderMusubi** [comment]
Yes, that is indeed helpful in finding the correct data to preview. However, a fallback is usually also implemented, by using the first image found in the page and Dublin Core is sometimes also used. See for example

    <head profile="http://a9.com/-/spec/opensearch/1.1/" prefix="og: http://ogp.me/ns#">
      ...
      <title>FOSDEM 2016 - Home</title>
      <link rel="schema.DC" href="http://purl.org/DC/elements/1.0/" />
      <meta name="DC.Title" content="Home"/>
      <meta name="DC.Format" content="text/html" />
      <meta name="DC.Language" content="en" />
      <meta name="DC.Subject" content="FOSDEM 2016" />
      <meta name="DC.Rights" content="http://creativecommons.org/licenses/by-sa/2.0/be/" />
      <meta property="og:title" content="FOSDEM 2016 - Home" />
      <meta property="og:type" content="website" />
      <meta property="og:image" content="https://fosdem.org/2016/assets/style/logo-gear-7204a6874eb0128932db10ff4030910401ac06f4e907f8b4a40da24ba592b252.png" />
    </head>

**Comment by jkufner** [comment]
This will be very useful for spammers. "click here to buy some... http://..." and Conversation clicks for you.

**Comment by PanderMusubi** [comment]
Perhaps it should be configurable per account to have more control over it. Personally I never have spammers on my chats or conferences.

**Comment by SecTec** [comment]
Hmm... In times of Stagefright when you only need an mp4 video to infect an Android device you probably don't want display any media by default.

**Comment by DoM1niC** [comment]
Whats About Latex??? https://modules.prosody.im/mod_latex.html

**Comment by d9h02f** [comment]
What kind of imbecile sends LaTeX through instant messaging?

**Comment by Natureshadow** [comment]
What about your language, eh?

And yes, LaTeX is very helpful for scientific conversations.

And above all, why should LaTeX be stranger than HTML?

**Comment by d9h02f** [comment]
I, too, enjoy typing LaTeX on a mobile phone. No better way to spend the day than to type endless backslashes and all sorts of brackets with your thumb. While we're at it, why not add LaTeX autocomplete as well? In fact Conversations should also recognise if the content of a message is written in a certain programming language and colour code everything. Maybe add an option to auto-execute as well, as long as it's an expert option nothing's wrong.

What you are suggesting is pure feature creep, completely useless bloat that only a few ultra nerds will use to show off. I don't even think providing previews of HTTP URLs is necessary but at least it could maybe be of some use sometime for someone.

LaTeX works relatively well for writing documents. Don't shoehorn it into things it has nothing to do with.

**Comment by ildar** [comment]
I suggest LaTeX voters to open a separate issue.
I vote for the opt-in preview feature with option: Off (default); Only in conversations with buddies in blist; In all conversations (dangerous, double check!).

**Comment by ildar** [comment]
Also I guess that showing image previews scaled down by Conversations may be considered quite safe (?). Though not quite sure.

**Comment by andypl** [comment]
Hi,

I have that feature in my SMS/MMS app (Textra) and I would gladly have it in Conversaionts app. 
If I send nor receive an url link, Textra read tag: 

`<meta property="og:description" content="">`

 and put it as website description and 

`<meta property="og:image" content="">`

 tag for image preview. It look really nice and its sometimes helpful because I know what is in url without opening it in browser ;) 

About privacy/security: This feature should be/or would be disabled for people who not receive our presence subscription, so if spammer send as link, then conversation would not display its preview :)

**Comment by ildar** [comment]
Privacy/Security: enable previews only for known contacts.
Security: also possible to use Adblock or Mozilla URL reputational DBs,
probably through a plugin.
What do you think?

**Comment by licaon-kter** [comment]
Firefox can be used as a webview so you'd gain any security loaded there.

But how do you use that? Load the page in background and save a screeenshot?

**Comment by ildar** [comment]
URL preview just shows a few meaningful words and the 1st picture downsized.
I don't see how FF can help here.

**Comment by licaon-kter** [comment]
@ildar to get that you'll need to parse something, if that metadata is not in HEAD you'll fallback to what?

**Comment by ildar** [comment]
Anyway that Pix-Art fork gone a lot further in this feature.

**Comment by licaon-kter** [comment]
@ildar Did you use that and found it useful? Working? Not annoying?

Maybe I'm the only one where the preview disappears at random and the servers are contacted every time I hit back on or open the chat.

**Comment by ildar** [comment]
I'm not saying how good it's implemented in P-A-M. It is offtopic here, in
Conversations tracker. Plz mention me at the corresponding issue at P-A-M
I'm just pointing users to alternative(s) in this stagnated issue.

**Comment by licaon-kter** [comment]
Maybe this helps: https://www.mysk.blog/2020/10/25/link-previews/

**Comment by wiktor-k** [comment]
This is a good overview but given Conversations design decisions (E2EE as first-class feature) this leaves out server generated previews ("Approach 3").

Now whichever option you pick the preview display should be strictly limited to "trusted" contacts. The same logic that affects file autodownload can be reused.

If the preview was to be generated on the sender side then this additionally requires full-stanza encryption.

**Comment by licaon-kter** [comment]
Blabber now allows server owners to host their own preview scripts 👍 
https://codeberg.org/kriztan/blabber.im/commit/8287e3e39e9aac7de28aebb00b93459057289f0b

**Comment by kriztan** [comment]
@licaon-kter this is not correct. This is only the server rendered map preview and not a http preview. Http previews are generated in the app.

**Comment by shodanx2** [comment]
Given the current state of development of conversation.im (functional and excellent), I think this feature of OGP content preview is currently the most dearly missed feature when compared with all other IM clients we used previously (sms,facebook,steam)

In response to previous discussion

I believe it should be enabled by default but possible to turn off for the few who would turn it off 

In a perfect world the OGP content request would be performed by the xmpp server or another trusted proxy, thus protecting the user's anonymity and hidden presence 

There should be a manual fallback preview method if OGP is not supported by the URL

It should be possible for the person who sends the URL, to push preview data so that the recipient doesn't have to get it for themselves (imagine posting a link to a 100 user chat room, it would be wasteful for have every client in that chatroom hammer the server with requests when the original sender could have bundled the title and thumbnail automatically with his URL)

If a URL has a title and thumbnail, omit or trunkate the displaying the URL ?

**Comment by licaon-kter** [comment]
Apparently not having previews is better for security :shrug: 

https://www.sevarg.net/2022/07/20/ios16-lockdown-mode-browser-analysis/
> Some features, like link previews, are disabled. GREAT!
> ...link previews are quite useful for an attacker. By sending a message to your phone, I might be able to make it reach out to a server I control, delivering a payload of whatever I want. Turning all this crap off is good.

**Comment by shodanx2** [comment]
If people need security, they can enable airplane mode.

Normal people want a preview before they click

Ok, you want a compromise ? Just ask the server to resolve the OGP preview for you. 
It will just run a sandboxed process that will download the image and text preview and send it back to you right back in the place where you received the URL. And the client should understand they are meant to visually substitute the raw URL with the received preview (and put the preview data in the history, so that it will remain visible after the link has died). 

This method is a great way to push the complexity on the server and also a way to implement fall back when OGP is not supported by the server. This will also protect the user from revealing his IP address and even presence status since the server will resolve the OGP even if the user is absent. 

This should be the default behaviour.

And, even all that is still trailing behind the competition because that's just text and images, users already expect to preview and view animation, sound files, video, 3d renders directly embedded from their chat application.

It's very hard to push to XMPP adoption when you tell the user, "yes, there's nobody on there except me, also a lot of feature you expect, just aren't supported yet so the experience will be a downgrade to what you already use". 

I have friends that I managed to convince to jump through all hoops and get an xmpp client on their phones. And they still send me messages with facebook messenger.

**Comment by licaon-kter** [comment]
@shodanx2 oh, so we should leak to the server our conversation links? Got it. I forgot what's "modern".

**Comment by shodanx2** [comment]
Default should be, your client asks the server "resolve this preview for me".

Don't trust the server ? Your client resolves it 

Don't trust the person you're talking with ? Don't resolve the previews. (but really, your computer should be completely impervious to URLs)

When you add a contact, by default there's an already checked option to share presence data with your contact. There could be an extra checkbox or combobox below that for how you want to handle previews and the default should be as above.

**Comment by shodanx2** [comment]
There is another possibility, the client of the person sending the URL, could push the preview data along so the receiver don't have to obtain it.
